### PR TITLE
feat(JMS): add JMS family (properties only)

### PR DESCRIPTION
**Pull request by RyanSkraba**
**Please check if the PR fulfills these requirements**

- [X] The commit message follows Talend standard
- [X] Tests for the changes have been added (for bug fixes / features)

**What is the current behavior?** (You can also link to an open issue here)

There is no component available for the JMS technology.

**What is the new behavior?**

Starting point for the design-time properties of the JMS datastore, dataset and input/output components.

**Does this PR introduce a breaking change?**

- [ ] Yes
- [ ] No

If this PR contains a breaking change, please describe the impact and migration path for existing applications: ...

Nobody depends on this component yet.

**Other information**:

The runtime work is in progress, so the contents of the jsm-runtime_1_1 module are works in progress -- we'd really like to get a review on the jms-definition at this point and include it in the REST API to be consumed by the UI (for design-only at this point).

### CODE REVIEW

**Comment by build-talend** [comment]
Pull request test result: [failure](https://ci-common.datapwn.com/job/components_pull-request_master/557/) [:x:](https://ci-common.datapwn.com/job/components_pull-request_master/557/)

File: components/components-jms/jms-definition/src/main/java/org/talend/components/jms/input/JmsInputDefinition.java [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+        super(COMPONENT_NAME);
+    }
+
+    @Override
+    public Class<? extends ComponentProperties> getPropertyClass() {
+        return JmsInputProperties.class;
+    }
+
+    public RuntimeInfo getRuntimeInfo(Properties properties, ConnectorTopology connectorTopology) {
+        return new SimpleRuntimeInfo(this.getClass().getClassLoader(),
```
Comment (sgandon) [inline components/components-jms/jms-definition/src/main/java/org/talend/components/jms/input/JmsInputDefinition.java]: I think you should be using JmsVersion here to switch the runtime no ?

File: components/components-jms/jms-runtime_1_1/src/main/java/org/talend/components/jms/runtime_1_1/JmsDatastoreRuntime.java [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+        Connection connection = connectionFactory.createConnection();
+        connection.start();
+        connection.close();
+
+        if (connection != null) {
+            return Arrays.asList(ValidationResult.OK);
+        }
+
+        } catch (JMSException e) {
+            e.printStackTrace();
```
Comment (sgandon) [inline components/components-jms/jms-runtime_1_1/src/main/java/org/talend/components/jms/runtime_1_1/JmsDatastoreRuntime.java]: no printStacjTrace please.

File: components/components-jms/jms-definition/src/main/java/org/talend/components/jms/JmsDatastoreProperties.java [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+//import javax.naming.NamingException;
+
+public class JmsDatastoreProperties extends PropertiesImpl implements DatastoreProperties {
+
+    public enum JmsVersion{
+        V_1_1,
+        V_2_0
+    }
+
+    public SchemaProperties main = new SchemaProperties("main");
```
Comment (sgandon) [inline components/components-jms/jms-definition/src/main/java/org/talend/components/jms/JmsDatastoreProperties.java]: no schema for datastore.

File: components/components-jms/jms-definition/src/main/java/org/talend/components/jms/JmsDatastoreProperties.java [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+
+    @Override
+    public void refreshLayout(Form form) {
+        super.refreshLayout(form);
+        // Main properties
+        if (form.getName().equals(Form.MAIN)) {
+                form.getWidget(version.getName()).setVisible();
+                form.getWidget(contextProvider.getName()).setVisible();
+                form.getWidget(serverUrl.getName()).setVisible();
+                if (needUserIdentity.getValue()) {
```
Comment (bchen-talend) [inline components/components-jms/jms-definition/src/main/java/org/talend/components/jms/JmsDatastoreProperties.java]: need a method named afterNeedUserIdentity to trigger

File: components/components-jms/jms-definition/src/main/java/org/talend/components/jms/JmsDatastoreProperties.java [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+                    form.getWidget(userPassword.getName()).setVisible();
+                } else {
+                    form.getWidget(userName.getName()).setHidden();
+                    form.getWidget(userPassword.getName()).setHidden();
+                }
+        }
+        // Advanced Properties
+        if (form.getName().equals(Form.ADVANCED)){
+            form.getWidget(useHttps.getName()).setVisible();
+            if (useHttps.getValue()){
```
Comment (bchen-talend) [inline components/components-jms/jms-definition/src/main/java/org/talend/components/jms/JmsDatastoreProperties.java]: need a method named afterUseHttps to trigger too.

File: components/components-jms/jms-definition/src/main/java/org/talend/components/jms/JmsDatastoreProperties.java [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+        V_2_0
+    }
+
+    public SchemaProperties main = new SchemaProperties("main");
+
+    public JmsDatastoreProperties(String name) {
+        super(name);
+    }
+
+    // FIXME Default values are not working => There are just some examples for the user
```
Comment (bchen-talend) [inline components/components-jms/jms-definition/src/main/java/org/talend/components/jms/JmsDatastoreProperties.java]: use setupProperties for set default value

File: components/components-jms/jms-definition/src/main/java/org/talend/components/jms/JmsDatastoreProperties.java [hunk 7]
```diff
@@ -0,0 +1,121 @@
...
+                if (needUserIdentity.getValue()) {
+                    form.getWidget(userName.getName()).setVisible();
+                    form.getWidget(userPassword.getName()).setVisible();
+                } else {
+                    form.getWidget(userName.getName()).setHidden();
+                    form.getWidget(userPassword.getName()).setHidden();
+                }
+        }
+        // Advanced Properties
+        if (form.getName().equals(Form.ADVANCED)){
```
Comment (bchen-talend) [inline components/components-jms/jms-definition/src/main/java/org/talend/components/jms/JmsDatastoreProperties.java]: just discuss, we need define advanced for datastore?

File: components/components-jms/jms-definition/src/main/java/org/talend/components/jms/JmsDatastoreProperties.java [hunk 8]
```diff
@@ -0,0 +1,121 @@
...
+    @Override
+    public void refreshLayout(Form form) {
+        super.refreshLayout(form);
+        // Main properties
+        if (form.getName().equals(Form.MAIN)) {
+                form.getWidget(version.getName()).setVisible();
+                form.getWidget(contextProvider.getName()).setVisible();
+                form.getWidget(serverUrl.getName()).setVisible();
+                if (needUserIdentity.getValue()) {
+                    form.getWidget(userName.getName()).setVisible();
```
Comment (bchen-talend) [inline components/components-jms/jms-definition/src/main/java/org/talend/components/jms/JmsDatastoreProperties.java]: thanks to pbailly, can use form.getWidget(userName).setVisible(needUserIdentity) directly

File: components/components-jms/jms-definition/src/main/java/org/talend/components/jms/JmsDatasetProperties.java [hunk 9]
```diff
@@ -0,0 +1,105 @@
...
+    @Override
+    public void setupLayout() {
+        Form mainForm = new Form(this, Form.MAIN);
+        mainForm.addRow(main.getForm(Form.MAIN));
+        mainForm.addRow(msgType);
+        mainForm.addRow(processingMode);
+    }
+
+    @Override
+    public void refreshLayout(Form form) {
```
Comment (bchen-talend) [inline components/components-jms/jms-definition/src/main/java/org/talend/components/jms/JmsDatasetProperties.java]: no need refresh if there is no trigger, the default is visible.

File: components/components-jms/jms-definition/src/main/java/org/talend/components/jms/input/JmsInputProperties.java [hunk 10]
```diff
@@ -0,0 +1,65 @@
...
+import static org.talend.daikon.properties.property.PropertyFactory.newInteger;
+import static org.talend.daikon.properties.property.PropertyFactory.newString;
+
+public class JmsInputProperties extends ComponentPropertiesImpl {
+
+    public JmsInputProperties(String name) {
+        super(name);
+    }
+
+    public SchemaProperties main = new SchemaProperties("main");
```
Comment (bchen-talend) [inline components/components-jms/jms-definition/src/main/java/org/talend/components/jms/input/JmsInputProperties.java]: has define the schema in JmsDatasetProperties already, so this one can be removed?

File: components/components-jms/jms-definition/src/main/java/org/talend/components/jms/input/JmsInputProperties.java [hunk 11]
```diff
@@ -0,0 +1,65 @@
...
+        Form mainForm = new Form(this, Form.MAIN);
+        mainForm.addRow(main);
+        mainForm.addRow(from);
+        mainForm.addRow(timeout);
+        mainForm.addRow(max_msg);
+        mainForm.addRow(msg_selector);
+    }
+
+    @Override
+    public void refreshLayout(Form form) {
```
Comment (bchen-talend) [inline components/components-jms/jms-definition/src/main/java/org/talend/components/jms/input/JmsInputProperties.java]: no need this method, same as dataset

File: components/components-jms/jms-definition/src/main/java/org/talend/components/jms/output/JmsOutputProperties.java [hunk 12]
```diff
@@ -0,0 +1,105 @@
...
+            form.getWidget(to.getName()).setHidden(false);
+        }
+        if (Form.ADVANCED.equals(form.getName())) {
+            form.getWidget(delivery_mode.getName()).setVisible();
+            form.getWidget(pool_max_total.getName()).setVisible();
+            form.getWidget(pool_max_wait.getName()).setVisible();
+            form.getWidget(pool_min_Idle.getName()).setVisible();
+            form.getWidget(pool_max_Idle.getName()).setVisible();
+            form.getWidget(pool_use_eviction.getName()).setVisible();
+            if (pool_use_eviction.getValue()) {
```
Comment (bchen-talend) [inline components/components-jms/jms-definition/src/main/java/org/talend/components/jms/output/JmsOutputProperties.java]: need after trigger method

File: components/components-jms/jms-definition/src/main/java/org/talend/components/jms/output/JmsOutputProperties.java [hunk 13]
```diff
@@ -0,0 +1,105 @@
...
+    public enum JmsAdvancedDeliveryMode {
+        Non_persistent,
+        persistent
+    }
+
+    public JmsOutputProperties(String name) {
+        super(name);
+    }
+
+    public SchemaProperties main = new SchemaProperties("main");
```
Comment (bchen-talend) [inline components/components-jms/jms-definition/src/main/java/org/talend/components/jms/output/JmsOutputProperties.java]: same as Input properties, no need for this one?

File: components/components-jms/jms-definition/src/main/java/org/talend/components/jms/JmsDatasetProperties.java [hunk 14]
```diff
@@ -0,0 +1,105 @@
...
+    public JmsDatasetProperties(String name) {
+        super(name);
+    }
+
+    public enum AdvancedPropertiesArrayType {
+        raw,
+        content
+    }
+
+    public SchemaProperties main = new SchemaProperties("main");
```
Comment (bchen-talend) [inline components/components-jms/jms-definition/src/main/java/org/talend/components/jms/JmsDatasetProperties.java]: no need it, as it in dataset

File: components/components-jms/jms-runtime_1_1/src/main/java/org/talend/components/jms/runtime_1_1/JmsOutputPTransformRuntime.java [hunk 15]
```diff
@@ -0,0 +1,77 @@
...
+        System.out.println("deb apply");
+        PCollection<IndexedRecord> test = objectPCollection.apply("ExtractIndexedRecord", ParDo.of(new DoFn<Object, IndexedRecord>() {
+            IndexedRecordConverter converter;
+            transient AvroRegistry avroRegistry = new AvroRegistry();
+            @DoFn.ProcessElement public void processElement(ProcessContext c) throws Exception {
+                if (c.element() == null){
+                    return;
+                }
+                if (converter == null){
+                    converter = avroRegistry.createIndexedRecordConverter(c.element().getClass());
```
Comment (bchen-talend) [inline components/components-jms/jms-runtime_1_1/src/main/java/org/talend/components/jms/runtime_1_1/JmsOutputPTransformRuntime.java]: question here is, how the AvroRegistry return properly converter by object, does it registry all the converter when init?

File: components/components-jms/jms-definition/src/main/java/org/talend/components/jms/JmsDatastoreDefinition.java [hunk 16]
```diff
@@ -0,0 +1,63 @@
...
+    }
+
+    @Override
+    public String getImagePath() {
+        return null;
+    }
+
+    @Override
+    public DatasetProperties createDatasetProperties(JmsDatastoreProperties storeProp) {
+        JmsDatasetProperties setProp = new JmsDatasetProperties(null);
```
Comment (sgandon) [inline components/components-jms/jms-definition/src/main/java/org/talend/components/jms/JmsDatastoreDefinition.java]: yes absolutely.

File: components/components-jms/jms-definition/src/main/java/org/talend/components/jms/input/JmsInputProperties.java (hunk 10 above)
Comment (sgandon) [inline components/components-jms/jms-definition/src/main/java/org/talend/components/jms/input/JmsInputProperties.java]: yes the schema is on the dataset. can you remove it from here ?

File: components/components-jms/jms-definition/src/main/java/org/talend/components/jms/input/JmsInputDefinition.java (hunk 1 above)
Comment (StefanLMNG) [inline components/components-jms/jms-definition/src/main/java/org/talend/components/jms/input/JmsInputDefinition.java]: For the moment we are only using one Runtime's version (1_1). But as soon as another version will be implemented, we will need to use a switch to use the runtime version.

**Comment by build-talend** [comment]
Pull request test result: [failure](https://ci-common.datapwn.com/job/components_pull-request_master/574/) [:x:](https://ci-common.datapwn.com/job/components_pull-request_master/574/)

**Comment by build-talend** [comment]
Pull request test result: [failure](https://ci-common.datapwn.com/job/components_pull-request_master/599/) [:x:](https://ci-common.datapwn.com/job/components_pull-request_master/599/)

**Pull request closed without merging**